### PR TITLE
Fix bugs found in full-repo review (search spaces, Pareto selection, GPU cleanup, SBC coverage)

### PR DESCRIPTION
**Pull request by matthiaskloft**
## Summary

A full-repo review (correctness, robustness, simplification) surfaced 8 issues, each fixed and verified in its own commit:

- **`TimeSeriesTransformerSpace` crash for `num_layers != 2`** — BayesFlow's `TimeSeriesTransformer` defaults `mlp_depths=(2, 2)`; the search space resized every other tuple to `num_layers` but not this one, so any sampled `num_layers != 2` raised a length-mismatch `ValueError`. Reproduced directly against the installed `bayesflow` package before fixing.
- **`select_best_trial` could return a Pareto-dominated trial** — the mean-rank tiebreak ranked over *all* Phase-1 survivors instead of the already-computed Pareto front, so a dominated candidate could out-rank every front member and get returned as "best". Added a regression test that constructs this exact scenario.
- **`MemoryError` skipped GPU cleanup** — every other failure path in `GenericObjective.__call__` calls `cleanup_trial()` (releases the CUDA cache) before returning/raising; the `MemoryError` branch alone didn't. It still propagates and halts the study as before (confirmed as intended) — it just no longer leaves the cache dirty for any caller that catches it upstream.
- **`denormalize_param_count` didn't round-trip with `normalize_param_count`** — the latter silently auto-tightens `min_count` when only `max_count` is overridden; the former had no equivalent, so composing the two didn't invert correctly. Extracted the shared logic into `_resolve_min_count`.
- **Checkpoint pool ranked by only the first Pareto objective** — `values[0]` was used as the ranking key even in multi-objective `"pareto"` mode, silently discarding checkpoints that were better on every other metric. Added `mean_objective_score()` (mean of all objectives excluding the trailing cost score) and reused it for both checkpoint ranking and `study.py`'s pre-existing `_mean_ranking_key`, which duplicated the same formula.
- **SBC coverage rank normalization was missing its continuity correction** — `validation/registry.py` computed `ranks / (n+1)` while an adjacent comment claimed a continuity correction was applied. The project's own reference notes (`docs/references/talts2018_sbc.md`) and the Talts et al. (2018) paper text confirm `(rank + 0.5) / (L + 1)` (Blom, 1958) as the documented formula, already correctly implemented in `sbc_tests.py`. Brought `registry.py` in line.
- **Dead `dimensions`/`sample` overrides** in `TrainingSpace`, `ConsistencyModelSpace`, `StableConsistencyModelSpace`, and `FlowMatchingSpace` — each reimplemented exactly what `BaseSearchSpace` already auto-derives from dataclass fields. Removed.
- **Boolean hyperparameters modeled as `CategoricalDimension`** — 8 fields across 5 files used the generic N-choices type for something fundamentally binary, with `choices=` never actually used since all were constants. Added a dedicated `BoolDimension` type (tunable over `{True, False}` when unset, mirroring the constant/tunable split of the other dimension types) and migrated these sites.

## Test plan

- [x] Full suite: `KERAS_BACKEND=torch pytest tests/ -q` → 722 passed
- [x] `ruff check src/ tests/` → all checks passed
- [x] Each bug reproduced/verified against the actual `bayesflow` package or project reference docs before fixing (not fixed from assumption)
- [x] Regression tests added for the `select_best_trial` Pareto bug and the checkpoint-pool ranking bug

Co-Authored-By: Claude Sonnet 5 <noreply@anthropic.com>

## Summary by CodeRabbit

* **New Features**
  * Added support for boolean hyperparameters across search spaces, including new fixed configuration for time-series transformer MLP depth.
  * Exposed additional public objective utilities and a boolean dimension type for easier customization.

* **Bug Fixes**
  * Improved multi-objective checkpointing and warm-start ranking by averaging all metric values (excluding the final cost component).
  * Ensured trials are cleaned up when memory limits trigger errors.
  * Refined “best trial” selection to break ties only among Pareto-front candidates.
  * Adjusted coverage metric normalization for more stable calibration.

* **Refactor**
  * Centralized parameter-count bound logic and standardized scoring/ranking helpers across the optimization flow.

### CODE REVIEW

File: src/bayesflow_hpo/results/extraction.py [hunk 1]
```diff
@@ -888,11 +890,16 @@ def select_best_trial(
...
+    # Pick the Pareto-front member with lowest mean rank, breaking ties
+    # by trial number for determinism.
+    front_mean_ranks = mean_ranks[front_indices]
+    min_rank = front_mean_ranks.min()
+    tied_front_indices = [
+        idx
+        for idx, rank in zip(front_indices, front_mean_ranks)
+        if rank == min_rank
+    ]
+    best_idx = min(tied_front_indices, key=lambda i: candidates[i].number)
```
[the remark this answers is not in the document]
Comment (matthiaskloft) [reply src/bayesflow_hpo/results/extraction.py]: Added explicit type hints to `front_mean_ranks`, `min_rank`, `tied_front_indices`, and `best_idx` in bcfcd1f.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: tests/test_results/test_extraction.py [hunk 2]
```diff
@@ -777,6 +777,46 @@ def test_pareto_front_contains_non_dominated(self):
...
+                values=[float(x), float(y)],
+                state=optuna.trial.TrialState.COMPLETE,
+            )
+            study.add_trial(trial)
+
+        trial, result = select_best_trial(study, priorities=[("m1", 1000.0)])
+
+        # Trial 0 is dominated and must never be selected.
+        assert trial.number != 0
+        assert trial.number in {t.number for t in result.pareto_front}
```
[the remark this answers is not in the document]
Comment (matthiaskloft) [reply tests/test_results/test_extraction.py]: Good catch — the original threshold (1000.0) let m1 drop out of Phase 2 entirely, so the test was only exercising a single-objective (m2) selection, which trivially avoids the bug. Switched to the suggested unmet threshold (0.0) so m1 is promoted and both objectives participate in bcfcd1f, with assertions matching the new selection.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: tests/test_search_spaces/test_flow_matching_space.py [hunk 3]
```diff
@@ -74,9 +74,22 @@ def test_time_embedding_dim_is_constant():
...
+    available, and validates the documented fallback behavior
+    (``"tsit5"``/``"adaptive"``) when it isn't — rather than crashing on
+    BayesFlow versions that no longer expose it.
+    """
     space = FlowMatchingSpace()
     timemlp_sig = inspect.signature(bf.networks.TimeMLP)
-    integrate_defaults = bf.networks.FlowMatching.INTEGRATE_DEFAULT_CONFIG
+    integrate_defaults = getattr(
+        bf.networks.FlowMatching, "INTEGRATE_DEFAULT_CONFIG", None
+    )
```
[the remark this answers is not in the document]
Comment (matthiaskloft) [reply tests/test_search_spaces/test_flow_matching_space.py]: Skipped — this isn't specific to this file. `KERAS_BACKEND` is set once at the CI job level (`.github/workflows/ci.yml:30`) and via the project's documented `export KERAS_BACKEND=torch` convention (CLAUDE.md) for local runs, the same as every other test file in this suite that imports `bayesflow_hpo`/`bayesflow` at module scope. Setting it only in this one file would be inconsistent with the rest of the test suite and wouldn't fix anything if the env var were actually missing (every test module would fail the same way, not just this one).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: tests/test_search_spaces/test_flow_matching_space.py [hunk 4]
```diff
@@ -94,8 +107,13 @@ def test_untuned_defaults_match_bayesflow_defaults():
...
     )
-    assert space.integrate_method.constant == integrate_defaults["method"]
-    assert space.integrate_steps.constant == integrate_defaults["steps"]
+
+    if isinstance(integrate_defaults, dict):
+        assert space.integrate_method.constant == integrate_defaults["method"]
+        assert space.integrate_steps.constant == integrate_defaults["steps"]
+    else:
+        assert space.integrate_method.constant == "tsit5"
+        assert space.integrate_steps.constant == "adaptive"
```
[the remark this answers is not in the document]
Comment (matthiaskloft) [reply tests/test_search_spaces/test_flow_matching_space.py]: Already resolved in 48d04ed (as CodeRabbit's own annotation notes) — that commit replaced the direct `integrate_defaults["method"]`/`["steps"]` indexing entirely with calls to the production `_flowmatching_integrate_default` helper, which does the `.get(param, fallback)` lookup internally. No further change needed.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
